test/lib/checkevents.go

```diff
@@ -235,8 +236,34 @@ func (ei *EventInfoStore) WaitAtLeastNMatch(f EventInfoMatchFunc, n int) ([]Even
 	return matchRet, internalErr
 }
 
+// Wait for at least minCount events with source exactly matching source and data contained within the event
+// data field.  If source is the empty string, don't check the source.  If maxCount is >0, return an error
+// if more than maxCount entries are seen.
+func (ei *EventInfoStore) WaitMatchSourceData(source string, data string, minCount int, maxCount int) error {
```

Comment:
Exactly like i sad in the comment above, more than these "static asserts" we could compose the matcher composing different predicates

Reply:
That should be doable... I'm actually doing something SLIGHTLY similar already with ValidEvFunc wrapper which converts the generic interface (that lets us detect deliveries that weren't successfully delivered) to a specific interface that only matches on validly decoded Events, so normal tests don't have to do a bunch of checks of if (event != nil)...  This would be a somewhat similar pattern.

Reply:
But the `ValidEvFunc` returns a bool, it would be cool to return an err to return a message that tells exactly what doesn't match. In that case we could have such methods directly in sdk-go like https://github.com/cloudevents/sdk-go/pull/522, so we use the matchers from sdk directly to compose the `ValidEvFunc`

Reply:
(We don't need to do it now, let me change the sdk-go #522 first)

Reply:
Look at this, does that makes sense to you? https://github.com/cloudevents/sdk-go/pull/522/files#diff-4b881640a24bb5860ddad74261b9c24e

Reply:
Yep.  That looks good.  There are a few case we want to go beyond that (looking at HTTP headers, or errors in reception), but I've already split the functions into separate interfaces for matching events and for matching the greater set of things around reception (since we very rarely want to do the latter and deal with the complexities of it)... I'm happy if we move the recordevents predicates to this.   Even if this isn't fully in yet, I can at least move EventMatchFunc to this signature (switch bool to err) to start making it compatible.

Reply:
> Even if this isn't fully in yet, I can at least move EventMatchFunc to this signature (switch bool to err) to start making it compatible.

Yes please do it, so then when we merge on sdk-go and update here, we can easily port it

Reply:
Just return a "dummy error" for now and we'll improve it later